### DISCUSSION TITLE
Is DGS web-aware framework?

### DISCUSSION THREAD

**Discussion by zenghou** [Netflix/dgs-framework#592] [Q&A]
I'm trying to use [scoped beans](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-factory-scopes) for each request, and it requires a web-aware `ApplicationContext`. In Spring Boot, I know having `RestControllers` would make the application web-aware. Does it work with data fetchers in DGS too? My guess is yes, but I'd like to be extra sure about this.

The use case:

```
@Component
@RequestScope
@NoArgsConstructor
@Getter
@Setter
public class RequestContextHolder {
    private RequestContext requestContext;
}
```

Setting up a custom context for each request to store the authorization details for the request that will be used for validation
```
    public RequestContext build(Map<String, ?> map, HttpHeaders httpHeaders, WebRequest webRequest) {
        RequestContext requestContext = new RequestContext();
        requestContext.storeAuthStuff();
        requestContextHolder.setRequestContext(requestContext);
        ...
    }
```

In the data fetcher method, I will be calling helper methods from other classes and I do not want to pass in the context as an extra parameter, which is why I am thinking of autowiring the `RequestContextHolder`

e.g.
```
dataFetcher(DataFetchingEnvironment dfe) {
    AnotherClass.helperMethod();   
}


public class AnotherClass {
   @Autowired
    RequestContextHolder ctxHolder;
    
   public String helperMethod() {
       RequestContext requestContext = ctxHolder.getRequestContext();
   }
    
}
```

**Comment by paulbakker** [comment]
It's hard to answer this with a simple yes or no, so I'll give some more context.

In the core of the framework, the only place we deal with anything we related is the following:
* Datafetcher arguments of `@RequestHeader`, `@CookieValue` and `@RequestParam`.
* The custom context builder that take a `WebRequest` as input. (or its equivalent in WebFlux)

If you're using WebMVC, the `/graphql` endpoint is a `@RestController`. When using Webflux it's a router function.
This means you _can_ use other features of the web stack (WebMVC/WebFlux) as well, such as filter.

One thing to be aware of is that query execution happens on different threads than the request itself, this is something that can cause issues when relying on anything thread local.

**Reply by zenghou** [reply]
Gotcha, thanks for the insight. I am thinking of just relying on the provided `SecurityContextHolder` to avoid working with `ThreadLocal`s directly. 

Since you mentioned that the query execution (I assume calling the `dataFetcher` method) will happen in its own thread, so it should be safe to get the `HttpServletRequest` from the `DataFetchingEnvironment` (instead of the `RequestContextHolder` which I guess is associated with the request itself), and set the `SecurityContext` in the thread for query execution.

For example: 

```
dataFetcher(DataFetchingEnvironment dfe) {
    reqHeaders = getHeadersFrom(dfe);
    // do auth stuff ...
    SecurityContextHolder.getContext().setAuthentication(auth);
    AnotherClass.helperMethod();  // in this helper method we will access the `SecurityContextHolder`
}
```

Would this be a safe approach?
